### DISCUSSION TITLE
How to configure multiple websites

### DISCUSSION THREAD

**Discussion by eduardo-g-silva** [markshust/docker-magento#456] [Q&A]
**This was writen by Marks (the guru) in a comment on his course and save my day...**

https://courses.m.academy/courses/487758/lectures/9064349

Change your docker-compose.yml file to:
...
services:
  app:
    build: images/nginx
    ...

Create a file at images/nginx/Dockerfile with the contents:

FROM markoshust/magento-nginx:1.18-3

COPY ./conf/default.conf /etc/nginx/conf.d/

Then create a file at images/nginx/conf/default.conf with the contents:

upstream fastcgi_backend {
  server unix:/sock/docker.sock;
}

map $http_host $MAGE_RUN_CODE {
    default base;
    domain1.test base;
    domain2.test other_website; # Put your secondary website code here
    # Add any other domains here using the same method
}

server {
  listen 8000;
  return 301 https://$host$request_uri;
}

server {
  listen 8443 ssl;

  ssl_certificate /etc/nginx/certs/nginx.crt;
  ssl_certificate_key /etc/nginx/certs/nginx.key;

  server_name
    domain1.test
    domain2.test
    # And any others...
  ;

  set $MAGE_ROOT /var/www/html;
  set $MAGE_RUN_TYPE website;

  include /var/www/html/nginx.conf;
}

You will also need to copy your src/nginx.conf.sample file over to src/nginx.conf, and update the contents of this block to include the MAGE_RUN_TYPE and MAGE_RUN_CODE variables:

...
location ~ ^/(index|get|static|errors/report|errors/404|errors/503|health_check)\.php$ {
    try_files $uri =404;
    fastcgi_pass   fastcgi_backend;
    fastcgi_buffers 1024 4k;

    fastcgi_param  PHP_FLAG  "session.auto_start=off \n suhosin.session.cryptua=off";
    fastcgi_param  PHP_VALUE "memory_limit=756M \n max_execution_time=18000";
    fastcgi_read_timeout 600s;
    fastcgi_connect_timeout 600s;

    fastcgi_index  index.php;
    fastcgi_param  SCRIPT_FILENAME  $document_root$fastcgi_script_name;
    fastcgi_param  MAGE_RUN_TYPE    $MAGE_RUN_TYPE;
    fastcgi_param  MAGE_RUN_CODE    $MAGE_RUN_CODE;
    include        fastcgi_params;

    fastcgi_buffer_size 128k;
}
...

Be sure to also update your docker-compose.dev.yml volume mounts if you are non-linux, so that your nginx reference looks at the new file:

...
services:
  app:
    volumes: &appvolumes
      # Host mounts with performance penalty, only put what is necessary here
      - ./src/app/code:/var/www/html/app/code:delegated
      ...
      - ./src/nginx.conf:/var/www/html/nginx.conf:delegated
...

Then, running bin/restart will rebuild the app service with the new Docker container. Hope this helps :)

**Comment by st3phani** [comment]
I think it would help if there was a configuration in the `env/magento.env` file.

**Reply by markshust** [reply]
I agree, something like this would be great. Perhaps a "multisites" param where we can comma-delimit the domains, and it would loop through and update/process the Nginx config. We may want to look into Nginx variables to do this, but I'm a bit out of the loop of how they work. It would also need to be backwards-compatible with single-domain sites.

**Comment by mgarciaebo** [comment]
(I deleted the previous message as I sent it from another account)

Hello,

I tried to follow several tutorials, including Mark's lesson and still can't run websites using the "folders approach".

Almost all the documentation available on the internet is focussed on subdomains: site1.domain.com/, site2.domain.com/, site3.domain.com/. What I'm trying to configure with no luck is: domain.com/site1/, domain.com/site2/, domain.com/site3/.

I'm not an expert in almost nothing, including Nginx :D however, I know that running sites using the folders configuration is a very common scenario and I wouldn't expect it's this difficult to both: find information about how to implement it using Nginx, and actually making it work.

Can anybody please point me out in the right direction, or give some hints?

Thank you!

**Reply by markshust** [reply]
I've never done this, and I don't think it's actually as common of a scenario as you think, as I've never come across this 😀. That said, you should be able to use a combination of the multi-sites configuration coupled with proxies. There's also a question if your root domain is a Magento instance itself, which could throw some hiccups into the mix. If not, you would create proxy directives for each path /site1, /site2, etc. and then those are basically treated as separate site configurations just like any other site. You'll also need the base URLs for each instance to match your desired URL setup.

**Comment by davemv** [comment]
Right now, Mark's tutorial works almost perfect, in my case i added this line inside Dockerfile, so it ends up like this:


FROM markoshust/magento-nginx:1.18-8
COPY --chown=app:app ./conf/default.conf /etc/nginx/conf.d/


Hope this helps someone.

**Reply by mitchkramez** [reply]
@davemv this saved us a bunch of time, and appears to be required with the current versions to get multisites working, thanks for sharing! without it, you'll get permissions errors saying that docker cannot modify the file for configuration.

**Comment by rubenromao** [comment]
Here's what worked for me for 3 websites.


1 - Stop the containers.


2 - Change "compose.yaml" nginx:
```
services:
  app:
#    image: markoshust/magento-nginx:1.18-8
    build: images/nginx
```


3 - Change "compose.dev.yaml" to include "nginx.conf":
```
#      - ./src/nginx.conf.sample:/var/www/html/nginx.conf:cached
      - ./src/nginx.conf:/var/www/html/nginx.conf:delegated
```


4 - Create "Doxkerfile" under "images/nginx/Dockerfile" with this content:
```
FROM markoshust/magento-nginx:1.18-8

COPY --chown=app:app ./conf/default.conf /etc/nginx/conf.d/
```


5 - Create "default.conf" under "images/nginx/conf/default.conf" and "Doxkerfile" :
```
upstream fastcgi_backend {
    server unix:/sock/docker.sock;
}

map $http_host $MAGE_RUN_CODE {
    default base;
    domain.local base;
    subdomain1.domain.local mageStoreCodeHere;
    subdomain2.domain.local mageStoreCodeHere;
}

server {
    listen 8000;
    return 301 https://$host$request_uri;
}

server {
    listen [::]:8443 ssl http2 ipv6only=on;
    listen 8443 ssl http2;

    ssl_certificate /etc/nginx/certs/nginx.crt;
    ssl_certificate_key /etc/nginx/certs/nginx.key;

    server_name domain.local subdomain1.domain.local subdomain2.domain.local;

    set $MAGE_ROOT /var/www/html;
    set $MAGE_RUN_TYPE website;

    fastcgi_buffer_size 64k;
    fastcgi_buffers 8 128k;

    include /var/www/html/nginx[.]conf;
}
```


6 - Change "nginx.conf" PHP entry like this:
```
# PHP entry point for main application
location ~ ^/(index|get|static|errors/report|errors/404|errors/503|health_check)\.php$ {
    try_files $uri =404;
    fastcgi_pass   fastcgi_backend;

    fastcgi_param  PHP_FLAG  "session.auto_start=off \n suhosin.session.cryptua=off";
    fastcgi_param  PHP_VALUE "memory_limit=756M \n max_execution_time=18000";
    fastcgi_read_timeout 600s;
    fastcgi_connect_timeout 600s;

    fastcgi_index  index.php;
    fastcgi_param  SCRIPT_FILENAME  $document_root$fastcgi_script_name;
    fastcgi_param MAGE_RUN_TYPE $MAGE_RUN_TYPE;
    fastcgi_param MAGE_RUN_CODE $MAGE_RUN_CODE;

    include        fastcgi_params;
}
```

7 - Start the containers "bin/start"
I strongly recommend to remove the existing nginx images volumes and containers before start.


8 - Setup the domains (I ran for all of them but tbh I don't know if it is needed)
`bin/setup-domain domain.local`
`bin/setup-domain subdomain1.domain.local`
`bin/setup-domain subdomain2.domain.local`


9 - Change your database URLs by replacing the base URLs for each website scope with your domains:
```bin/n98-magerun2 db:query "UPDATE `core_config_data` SET `value` = REPLACE(`value`, 'website1, 'domain.local')"```
```bin/n98-magerun2 db:query "UPDATE `core_config_data` SET `value` = REPLACE(`value`, 'website2, 'subdomain1.domain.local')"```
```bin/n98-magerun2 db:query "UPDATE `core_config_data` SET `value` = REPLACE(`value`, 'website3, 'subdomain2.domain.local')"```

10 - Flush the cache
`bin/magento c:f`

That's it.
I think I haven't miss anything.

Cheers

**Reply by pushpa-01** [reply]
Hello

I have followed above steps for add multistore domain.after that I have restart all services with command bin/restart and look like command work fine. but when I check with command "docker ps" then app service is not there. after that I run bin/setup-domain command for adding domain but getting error that service "app" is not running. can anyone has faced this issue?
I am getting below error
https://nimb.ws/kicbL0x

**Reply by rubenromao** [reply]
You should check the logs using `docker-compose logs` to understand why the service isn't running.

**Reply by pushpa-01** [reply]
I checked logs but I haven't found any log regarding "app" service. Also when I run bin/restart command it does not show any error.
https://i.imgur.com/u8QwCEa.png

It is working fine with "image: markoshust/magento-nginx:1.24-0" but not working with "build: images/nginx" in "compose.yaml" file.

**Reply by rubenromao** [reply]
You have something in your docker logs... 
Did you tried this command:
```
magento-vanilla-2.4.7$ docker-compose logs | grep app-1
app-1          | 2024/05/22 14:06:34 [info] 30#30: *462 client 127.0.0.1 closed keepalive connection
app-1          | 127.0.0.1 - - [22/May/2024:14:07:04 +0000] "GET / HTTP/1.1" 301 169 "-" "curl/8.5.0" "-"
app-1          | 2024/05/22 14:07:04 [info] 30#30: *463 client 127.0.0.1 closed keepalive connection
app-1          | 127.0.0.1 - - [22/May/2024:14:07:34 +0000] "GET / HTTP/1.1" 301 169 "-" "curl/8.5.0" "-"
app-1          | 2024/05/22 14:07:34 [info] 30#30: *464 client 127.0.0.1 closed keepalive connection
app-1          | 2024/05/22 14:08:04 [info] 30#30: *465 client 127.0.0.1 closed keepalive connection
app-1          | 127.0.0.1 - - [22/May/2024:14:08:04 +0000] "GET / HTTP/1.1" 301 169 "-" "curl/8.5.0" "-"
```
It should show you all app log entries.
Unless you've changed the service name from `app` to something else in the `docker-compose.yml` or `compose.yml` file, you should see something in the logs.

**Reply by pushpa-01** [reply]
yes.I found the issue and now container running fine.
Thank you for your help !!

**Reply by bhushanmeetanshi** [reply]
@rubenromao i have tried this steps but not helping me. is any other way?

**Reply by rubenromao** [reply]
Not that I'm aware of.
I can't help you without more detailed information like your log errors if any, your nginx default.conf file, etc.

**Reply by bhushanmeetanshi** [reply]
i have set all the thing that you have mentioned in steps but my second website foo.test is redirected to main website url

**Reply by bhushanmeetanshi** [reply]
i have slightly confused in db query steps

**Reply by bhushanmeetanshi** [reply]
@rubenromao i have executed all the steps but my new foo.test website not working it redirected to magento.test website. can you helpme if i miss any thing.

**Reply by pushpa-01** [reply]
> i have set all the thing that you have mentioned in steps but my second website foo.test is redirected to main website url

@bhushanmeetanshi have you run bin/setup-domain command for foo.test website also?

**Reply by bhushanmeetanshi** [reply]
> > i have set all the thing that you have mentioned in steps but my second website foo.test is redirected to main website url
> 
> @bhushanmeetanshi have you run bin/setup-domain command for foo.test website also?

Yes i have execute the same  command but its redirected to same main website magento.test instead of foo.test

**Reply by bhushanmeetanshi** [reply]
> > i have set all the thing that you have mentioned in steps but my second website foo.test is redirected to main website url
> 
> @bhushanmeetanshi have you run bin/setup-domain command for foo.test website also?

is any other process step require?

**Reply by pushpa-01** [reply]
> > > i have set all the thing that you have mentioned in steps but my second website foo.test is redirected to main website url
> > 
> > 
> > @bhushanmeetanshi have you run bin/setup-domain command for foo.test website also?
> 
> is any other process step require?

Have you facing in error while start docker process or any docker logs?

**Reply by bhushanmeetanshi** [reply]
> > > > i have set all the thing that you have mentioned in steps but my second website foo.test is redirected to main website url
> > > 
> > > 
> > > @bhushanmeetanshi have you run bin/setup-domain command for foo.test website also?
> > 
> > 
> > is any other process step require?
> 
> Have you facing in error while start docker process or any docker logs?

no docker is working fine. but newly created website not opening, thats the only problem

**Reply by rubenromao** [reply]
Did you check the store code name for your `foo.test` store and put it in nginx in `images/nginx/conf/default.conf`?
example:
```
map $http_host $MAGE_RUN_CODE {
    default base;
    domain.local base;
    subdomain1.domain.local mageStoreCodeHere;
    subdomain2.domain.local mageStoreCodeHere;
}
```

Also, **note** that I have this working using subdomais not doamais I'm not sure if that can be your issue too tbh.
Give it a try:
1st check your store codes in nginx are correct.
2nd try using subdomain.

Cheers

**Reply by bhushanmeetanshi** [reply]
Cheers working your solution now..

but 1 question is that


1. COPY --chown=app:app ./conf/default.conf /etc/nginx/conf.d/ (not work)
2. COPY ./conf/default.conf /etc/nginx/conf.d/ (work)

why this happens? whats the difference?

@rubenromao can u please guide me for that?

**Reply by rubenromao** [reply]
Is it working ?

**Reply by rubenromao** [reply]
the file user and group should be app:app and only the root user can do that.
Not sure how I achieve that... 
Do it that way and then go into the app container as root and change the ownership of the file if it isn't app:app you will need to run `chown app:app /etc/nginx/conf.d/default.conf`

**Reply by bhushanmeetanshi** [reply]
yes its working with this first approach of dockerfile

1 `COPY ./conf/default.conf /etc/nginx/conf.d/ (work)`
2. `COPY --chown=app:app ./conf/default.conf /etc/nginx/conf.d/default.conf (not work)`

i don't know whats difference between this 2 different approach. thats why im asking

**Reply by bhushanmeetanshi** [reply]
finally add your approach and restart containers and see that your approach also worked..

Thanks....

**Reply by rubenromao** [reply]
Glad it helped mate!
Cheers
